Improve airq handling of DeviceInfo

Sibgatulin opened the pull request:
Implement changes, [proposed](https://github.com/home-assistant/core/pull/90192#discussion_r1147186853) by @epenet in another PR.

Instead of fetching device information from the device once in `ConfigFlow` and then piping it through in `ConfigEntry.data`, only use as much as needed in `ConfigFlow.async_step_user`, then fetch again in `AirQCoordinator._async_update_data` if a key is missing.

Additionally, factor `AirQCoordinator` out into a sumbodule.

Add a simple test for `AirQCoordinator.device_info` update.

Positive side effect: `AirQCoordinator.device_info` is updated explicitly, instead of dumping the entire content of (a fully compatible) `TypedDict`, retrieved from `aioairq`.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Sibgatulin committed:
Reduce data sharing between ConfigFlow and DataUpdateCoordinator

Instead of fetching device information from the device once in
`ConfigFlow` and then piping it through in `ConfigEntry.data`,
only use as much as needed in `ConfigFlow.async_step_user`, then fetch
again in `AirQCoordinator._async_update_data` if a key is missing.

Additionally, factor `AirQCoordinator` out into a sumbodule.

Add a simple test for `AirQCoordinator.device_info` update.

Positive side effect: `AirQCoordinator.device_info` is
updated explicitly, instead of dumping the entire content of (a fully
compatible) `TypedDict`, retrieved from `aioairq`.

epenet commented on tests/components/airq/test_config_flow.py:
```diff
@@ -94,3 +93,21 @@ async def test_form_invalid_input(hass: HomeAssistant) -> None:
     assert result2["errors"] == {"base": "invalid_input"}
+
+
+async def test_duplicate_error(hass: HomeAssistant) -> None:
+    """Test that errors are shown when duplicates are added."""
```
Please remove this and keep it inside #90192

Sibgatulin replied:
This slipped through by my oversight, my apologies

epenet commented on tests/components/airq/test_coordinator.py:
```diff
@@ -0,0 +1,47 @@
+"""Test the air-Q data update coordinator."""
```
You shouldn't be testing the coordinator as a standalone.
It should instead be part of a full test suite.

You would have a couple of tests in `test_init.py`:
- One for successful initialisation + unload (eg `test_setup_config_and_unload`)
- One for failed initialisation (eg `test_async_setup_entry_not_ready`)

And you would have a platform test `test_sensor.py`:
- Ensure that device_registry is correctly populated
- Ensure that entity_registry is correctly populated
- Ensure that state_machine is correctly populated

`sfr_box` tests were recently refactored to use `SnapshotAssertion` so I think it's a good base.

Anyway this is probably a big chunk for a follow-up PR, and I recommend that you remove this test file for now.

Sibgatulin replied:
Thanks for the advice, I'll look into it. Removed the file for now.

epenet replied:
You can add coordinator.py to coverage ignore file for now.

Sibgatulin replied:
Of course, thanks

Sibgatulin committed:
Remove tests ill-suited to this PR

`test_config_flow.test_duplicate_error` slipped through by mistake,
while `test_coordinator.test_fetch_device_info_on_first_update` may need
a more thoroughly suite of accompanying tests

Sibgatulin committed:
Ignore airq/coordinator.py

...newly separated from airq/__init__.py, that's already in this list

epenet commented on .coveragerc:
```diff
@@ -37,6 +37,7 @@ omit =
     homeassistant/components/airnow/sensor.py
     homeassistant/components/airq/__init__.py
     homeassistant/components/airq/sensor.py
+    homeassistant/components/airq/coordinator.py
```
alphabetical order

Sibgatulin replied:
Sorry, fixed

Sibgatulin committed:
Reorder files alphabetically

Sibgatulin commented on tests/components/airq/test_config_flow.py:
```diff
@@ -48,7 +45,7 @@ async def test_form(hass: HomeAssistant) -> None:
     assert result2["type"] == FlowResultType.CREATE_ENTRY
     assert result2["title"] == TEST_DEVICE_INFO["name"]
-    assert result2["data"] == TEST_DATA_OUT
+    assert result2["data"] == TEST_USER_DATA
```
Thanks for the suggestion. In this choice, I think I'd opt for a `SnapshotAssertion` to avoid replication dict content. (Although, I feel a bit reluctant to lose the explicit assertion of the few fields of the `FlowResult` which I expect to be changed by the tested functionality.)

Should it be a part of this PR?

epenet replied:
IMO Only the "data" key should be compared against the snapshot.
Other fields are OK as they are.
If can be done in another PR.

epenet reviewed (approved):
LGTM 👍

epenet commented:
I think this can be set back to "ready for review", unless you are planning more changes?
